sourcegraph/cody
# remove `ContextFile.fileName`, use URIs instead

This PR addresses the following problems:

* `ContextFile.fileName` is overloaded for many different things, including: filenames (with different transformations applied), symbol names, and representations of context files in the chat input. This was making the code hard to read/understand/fix for me.
* URIs were encoding relative file paths (that were looked up at resolution-time in an ambiguous and duplicative manner), ranges, symbol names, etc. We should not try to jam this all into a URI; where needed, we should just pass these values in separate arguments or properties.

See https://sourcegraph.slack.com/archives/C05AGQYD528/p1704869760165279 for more info.

## Solution

* Better typing for `ContextFile` 
* Removed of `ContextFile.fileName`; use `ContextFile.uri` instead
* Removed a lot of unnecessary usage of `vscode.Uri.file` and `uri.fsPath`
* Removed "janky" usage of URIs to pass additional information, instead passing that in separate args or properties
* URIs are made into "nice" workspace-relative paths to show the user at *display time* (as late as possible) using `displayPath(uri)`, so that the data we pass around need not involve presentation concerns

## Notes

In some areas (such as mapping embeddings results filenames to URIs), support for multi-root workspaces regressed. I am not sure if it ever *actually* worked on multi-root workspaces. There were a lot of implicit relative file lookups that probably would find the wrong file for ambiguous filenames among the different workspace folders.

I'd like to merge this sooner and address those in a followup soon.

## Test plan

* CI
* Add `@`-mentioned files and symbols and ensure chat messages are sent and responses received properly. Also ensure that the chat is revived properly.
* Run NLS search and ensure the results are clickable.
